zip-0221.rst

```diff
@@ -0,0 +1,585 @@
+
+        # Merge peaks from right to left. 
+        # This will produce a list of peaks in reverse order
+        current = leaf
+        for peak in peaks[::-1]:
+            current_leaves = current.latest_height - current.earliest_height + 1
+            peak_leaves = peak.latest_height - peak.earliest_height + 1
+
+            if current_leaves == peak_leaves:
+                current = make_parent(peak, current)
+            else:
+                merged.append(current)
+                current = peak
+
+        # finally, bag the merged peaks
+        return bag_peaks(merged[::-1])
+
+In case of a block reorg, we have to delete the latest (i.e. rightmost) MMR leaf nodes, up to the reorg length. This operation is ``O(log(k))`` where ``k`` is the number of leaves in the right subtree of the MMR root.
+
+.. code-block:: python
+
+    def delete(root: ZcashMMRNode) -> ZcashMMRNode:
+        '''
+        Delete the rightmost leaf node from an existing MMR
+        Return the new tree root
+        '''
+
+        n_leaves = root.latest_height - root.earliest_height + 1
+        # if there were an odd number of leaves,
+        # simply replace root with left_child
+        if n_leaves & 1:
+            return root.left_child
+
+        # otherwise, we need to re-bag the peaks.
+        else:
+            # first peak
+            peaks = [root.left_child]
+
+            # we do this traversing the right (unbalanced) side of the tree
+            # we keep the left side (balanced subtree or leaf) of each subtree
+            # until we reach a leaf
+            subtree_root = root.right_child
+            while subtree_root.left_child:
+                peaks.push(tmp_root.left_child)
+                subtree_root = tmp_root.right_child
+
+        new_root = bag_peaks(peaks)
+        return new_root
+
+    def bag_peaks(peaks: List[ZcashMMRNode]) -> ZcashMMRNode:
+        '''
+        "Bag" a list of peaks, and return the final root
+        '''
+        root = peaks[0]
+        for i in range(1, len(peaks)):
+            root = make_parent(root, peaks[i])
+        return root
+
+Header modifications specification
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This ZIP introduces a new header version. It is identical to the current v4 header [#zcashBlock]_, except for the following changes:
+
+1. The version number is changed to `5`.
+2. ``hashSaplingFinalRoot`` is replaced by ``hashChainHistoryRoot``, as described above.
+
+The new block header format is:
+
+.. code-block:: C
+
+    class CBlockHeader
+    {
+    public:
+        // header
+        static const size_t HEADER_SIZE=4+32+32+32+4+4+32; // excluding Equihash solution
+        static const int32_t CURRENT_VERSION=5;
+        int32_t nVersion;
+        uint256 hashPrevBlock;
+        uint256 hashMerkleRoot;
+        uint256 hashChainHistoryRoot;
+        uint32_t nTime;
+        uint32_t nBits;
+        uint256 nNonce;
+        std::vector<unsigned char> nSolution;
+        ...
+    }
+
```

Comment:
It probably goes without saying, but `hashChainHistoryRoot` needs to be included in the PoW (which just requires another small change to CEquihashInput).

Reply:
The way this section is now phrased, this should be obvious.